### ISSUE TITLE
[contrib/ansible] Failure to provision a working kube cluster due to unresolveable hostnames

### ISSUE THREAD

**Issue by dcwangmit01** [kubernetes/contrib#1129]
I'm having trouble getting the most recent contrib/ansible scripts to successfully provision a Kubernetes cluster.  It used to work many days ago around kubernetes/contrib master branch commit 3634e300, but now fails at commit f48ce77f.  I'm hoping someone can provide advice that may allow me to use the unmodified stock code.  Otherwise I'm looking for approval of a proposed change which I can turn into a PR.

At some point recently, the contrib/ansible scripts seemed to have changed so that host arguments to processes are now hostnames instead of IPs.  At kubernetes/contrib commit 3634e300 the arguments to flanneld, kube-proxy, and kubelet were IP-based, and now they are hostname based.  This is also true of the contents of the kube config file found on the master at: ```/etc/kubernetes/kubectl.kubeconfig```

For example:
```
# before around master branch commit 3634e300
/usr/bin/flanneld -etcd-endpoints=http://{ip-address-of-the-master}:2379 ...
# now around master branch commit f48ce77f 
/usr/bin/flanneld -etcd-endpoints=http://{hostname-of-the-master}:2379 ...
```
Because the hostnames of cluster peers are unresolveable, the Kubernetes cluster fails to function.

In order to work around the problem, I created additional ansible tasks to update the /etc/hosts file of every node so that hostnames may resolve into IP addresses.  This solves the problem, but with the drawback of slowing down the ansible deploy due to the gathering of facts.

```
diff --git a/ansible/roles/pre-ansible/tasks/main.yml b/ansible/roles/pre-ansible/tasks/main.yml
index 9b21de4..21138b2 100644
--- a/ansible/roles/pre-ansible/tasks/main.yml
+++ b/ansible/roles/pre-ansible/tasks/main.yml
@@ -39,3 +39,22 @@

 - include: fedora-dnf.yml
   when: os_version.stdout|int >= 22 and 'Fedora' in distro.stdout and is_atomic is not defined
+
+- name: Gather cluster networking facts for /etc/hosts
+  setup:
+  register: networking_facts
+  tags:
+    - "/etc/hosts"
+
+- name: Insert/update /etc/hosts using ansible cluster facts
+  blockinfile:
+    dest: /etc/hosts
+    block: |
+      {% for host in groups['all'] %}
+        {{- hostvars[host].ansible_default_ipv4.address }} {{ host }}
+      {% endfor %}
+    marker: "# {mark} ANSIBLE MANAGED BLOCK"
+    create: no
+  when: networking_facts
+  tags:
+    - "/etc/hosts"
```
Tested on:

```
Kube Masters and Nodes: CentOS Linux release 7.2.1511 (Core)
Dev Machine: ansible 2.1.0.0
```

This is my first time touching Ansible.  Feedback is appreciated.  Thanks!

**Comment by ingvagabund** [comment]
Basically, you need what Vagrant does [1] with ``lineinfile`` module [2] before it runs the ansible deploy cluster playbook.

In my CI environment, all host names equals to their private IPs (automatically done by provisioning system). So there is no need for DNS resolution.

Other option is to use dynamic inventory. You need two pieces:

* a file (e.g. ``CLUSTER_RESOURCE``) with a list of IPs of your cluster. E.g.
```sh
MASTERS=10.8.168.107
ETCD=10.8.168.107
NODES=10.8.168.1,10.8.168.106
```

* script (e.g. python) that will read the file and generate a json file, e.g.:
```json
{
	"masters": {
		"hosts": ["10.8.168.107"],
		"vars": {
			"ansible_ssh_user": "root",
			"ansible_sudo": false
		}
	},
	"etcd": {
		"hosts": ["10.8.168.107"],
		"vars": {
			"ansible_ssh_user": "root",
			"ansible_sudo": false
		}
	},
	"nodes": {
		"hosts": ["10.8.168.1", "10.8.168.106"],
		"vars": {
			"ansible_ssh_user": "root",
			"ansible_sudo": false
		}
	}
}
```

I would prefer using the dynamic inventory if you run the ansible over dynamically provisioned cluster.

[1] https://github.com/kubernetes/contrib/blob/master/ansible/vagrant/vagrant-ansible.yml#L8
[2] http://docs.ansible.com/ansible/lineinfile_module.html

**Comment by ingvagabund** [comment]
@dcwangmit01 would the dynamic inventory help in your case?

**Comment by dcwangmit01** [comment]
@ingvagabund Thanks for responding, and for the detailed pointer in the right direction.  I'm going to take the dynamic inventory approach and write a script that interprets my ~/.ssh/config.

I also realize now that the unresolvable hostname issue I experienced was not due to code changes, but was instead due to changing my inventory to be based on hostnames defined in my ~/.ssh/config, instead of IP addresses.  Thus, the need to generate /etc/hosts (much like what you pointed out regarding vagrant-ansible.yml).

Closing this issue.
